### ISSUE TITLE
slate-history does not correctly undo selection movement

### ISSUE THREAD

**Issue by TheSpyder** [ianstormtaylor/slate#3756]
#### Do you want to request a _feature_ or report a _bug_?
bug

#### What's the current behavior?

After performing actions that modify a ranged selection, undo does not restore the selection correctly. In my replication case, selecting across paragraphs and then deleting does not undo correctly. The replication instructions are in the sandbox (and recorded in the gif).
https://codesandbox.io/s/undo-selection-bug-replication-case-wtkbr
![undo selection bug](https://user-images.githubusercontent.com/298292/86175375-8f181580-bb66-11ea-9195-926cdecb5584.gif)
Slate: 0.58.3
Browser: Chrome / Safari / Firefox
OS: Mac

#### What's the expected behavior?

Undo should restore the cursor exactly. I found a workaround - more on that along with a codesandbox link after the gif.
![undo selection expected behaviour](https://user-images.githubusercontent.com/298292/86175453-b969d300-bb66-11ea-8235-2196450938e2.gif)

The problem is that in scenarios such as this delete, Slate generates partial `set_selection` operations (similar to the issue raised in #3705). These are stateful operations making them very sensitive to the order they are applied.

[My workaround sandbox](https://codesandbox.io/s/undo-selection-bug-workaround-pjsf0) logs the order of operations in the history batch, and it's quite weird:
```
[
  {"type":"remove_text","path":[0,0],"offset":5,"text":" line"},
  {"type":"remove_text","path":[1,0],"offset":0,"text":"seco"},
  {"type":"merge_node","path":[1],"position":1,"target":null,"properties":{}},
  {"type":"set_selection","properties":{"anchor":{"path":[0,0],"offset":5}},"newProperties":{"anchor":{"path":[0,1],"offset":0}}},
  {"type":"merge_node","path":[0,1],"position":5,"target":null,"properties":{}}
]
```
Why is the `set_selection` in between the two `merge_node` operations? Perhaps only Ian knows for sure. Regardless, when undo is performed - which inverts these operations and applies them in reverse order - the partial selection update doesn't work. The actual work performed during the undo is:
- split the text node at `[0,0]` offset `5`
- move the anchor back to the end of the split text node (focus remains at the start of the new text node created by the split)
- split the block at `[0]` putting one text node in each block
- insert the removed text in the second block
- insert the removed text in the first block

I guess the assumption is that inserting the removed text in the first block will happen inside the cursor range, but it doesn't. It's inserted before the anchor which means the anchor remains at the end of the first paragraph.

My workaround moves the selection change to the _start_ of the batch. The effect of this is that the cursor is allowed to shift to the correct focus point (that was always fine) and then _after_ the removed text has been inserted it moves the anchor to the middle of the first line.

I have two thoughts about this:
1) the selection should collapse before any changes happen, not in the middle of the nodes merging (my workaround effectively applies this change after the fact)
2) the selection operation should set both anchor and focus, otherwise undo assumes the cursor hasn't moved, which is bad. This would fix #3705.

Fixing the second issue would also resolve a further issue with selection after undo. `slate-history` is attempting to compensate for stateful selection operations by restoring the cursor to where it was before reversing the undo, but it only keeps the most recent selection change so it's fairly broken as well. Here's another gif showing what happens if I move the cursor in between delete and undo. I leave the cursor after the final `n`, undo moves it one character back, but it should be moving to where it was after the delete action.
![Selection end wrong after undo with workaround](https://user-images.githubusercontent.com/298292/86186594-d6110580-bb7c-11ea-8666-c41b28f238e4.gif)
This doesn't happen in my implementation due to leveraging some OT smarts, but I thought I would mention it as well since it ties into the whole mess.

**Comment by cprodescu** [comment]
Thank you for filing this with comprehensive description!
I found this issue because the same bug manifests with inline nodes (links) and can be repro'd in the Slate Links example: https://www.slatejs.org/examples/links

Repro:
- Select first row
![Screen Shot 2020-08-14 at 4 31 42 PM](https://user-images.githubusercontent.com/1377013/90299702-f4407580-de4b-11ea-81a8-1a0cff5f96a0.png)
- Press backspace (deleting the row)
![Screen Shot 2020-08-14 at 4 34 22 PM](https://user-images.githubusercontent.com/1377013/90299713-07534580-de4c-11ea-9598-4ffcc030a582.png)
- Undo (ctr/cmd+z)
![Screen Shot 2020-08-14 at 4 31 52 PM](https://user-images.githubusercontent.com/1377013/90299724-0d492680-de4c-11ea-83d9-c6a635e6610a.png)

Expected:
- Row is fully selected
Actual:
- Only link is selected

**Comment by TheSpyder** [comment]
> Why is the `set_selection` in between the two `merge_node` operations? Perhaps only Ian knows for sure.

I've gained a lot of Slate knowledge in the last 5 months, and I can now answer my own question. This tends to happen when a core Slate action does not generate a valid document and built-in normalization fixes it. That unfortunately means the solution is probably not easy (my workaround is now causing a bug).

**Comment by bryanph** [comment]
In some cases, `set_selection` is a derived operation. When you call an `insert_text` operation you are also implicitly moving the selection forward by 1 (offset).

I kind of wonder whether it is possible to achieve the right behavior by not storing the `set_selection` ops in the history batch at all but instead "calculate" the resulting selection after undo or redo. 

So after inserting text/nodes and calling `undo()`: 
* The undo batch contains a "delete batch"
* After applying this batch, set the cursor to the position of the last delete `op`

After deleting text/nodes and calling `undo()`:
* The undo batch is an "insert batch"
* After applying this batch, set the cursor to the inserted range (if multiple `insert_node`, `insert_text` ops, need to add them all up somehow)

**Comment by TheSpyder** [comment]
That's a nice theory, but the selection stored in the undo history isn't always a collapsed selection. For example after applying an inline format, undo retains the range selection. And in that case, the format action might take multiple operations so there is no easy way to calculate what the selection should be after undo.

**Comment by bryanph** [comment]
@TheSpyder What I'm suggesting is that perhaps we can get by without storing the selection in the undo history in the first place. So for example in your example of applying an inline format, on undo just always select the formatted (or unformatted) text as part of the logic in `slate-history`.

**Comment by TheSpyder** [comment]
Yes, I understood your intent, but it's not that easy. Applying a format generally involves a minimum of two `split_node` operations and one `set_node`. Depending on how much content is selected it could get a lot more complicated than that.

Determining the start and end points of the selection reliably seems quite complex and IMO will cause more bugs than storing the selection does.

**Comment by bryanph** [comment]
@TheSpyder You're right about it being quite complex to determine the selection state given the ops. I tried an approach that is more similar to what you're doing here I suppose but integrate that explicitly into `slate-history`. See https://github.com/ianstormtaylor/slate/pull/4717

**Comment by TheSpyder** [comment]
It's a similar idea to mine, but yours is done in probably a better way 🤔

**Comment by miquik** [comment]
Hi, are there any news about this issue? I fine same problem and I would like to know If somebody find a workaround at least
Thanks

**Comment by bryanph** [comment]
@miquik https://github.com/ianstormtaylor/slate/pull/4717 might have fixed this for you
